File: examples/IBPD/explicit/ex1/generate_shell2d.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <set>
#include <functional>
#include <utility>
#include <cmath>
#include <cassert>

typedef std::pair<int, int> Edge;
typedef std::pair<double, double> EdgeProp;
struct EdgeComp : public std::binary_function<Edge, Edge, bool>
{
    inline bool operator()(const Edge& e1, const Edge& e2) const
    {
        return (e1.first < e2.first) || (e1.first == e2.first && e1.second < e2.second);
    }
};

void
sort_edge(Edge& e)
{
    if (e.first > e.second)
    {
        int temp = e.second;
        e.second = e.first;
        e.first = temp;
    }
    return;
}

int
main(int argc, char** argv)
{
    // Problem parameters
    const double R = 0.25;
    const double w = 0.0625;
    const double gamma = 0.1;
    const double x_c = 0.5;
    const double y_c = 0.5;

    const int N = 64;
    const double dx = 1.0 / (double)N;
    const double MFAC = 0.5;
    const double ds = MFAC * dx;

    const int n_x = ceil(2.0 * M_PI * R / ds);
    const int n_y = ceil(w / ds);
    const int totnode = n_x * n_y;

    const double area = ds * ds;  // cross-sectional area
    const double vol = area * ds; // volume of a material point
    const double delta = 1.015 * ds;
    const double scr0 = 50 * ds; // critical stretch.

    // Initialize vertices
    std::vector<std::vector<double> > coord(totnode, std::vector<double>(2));
    int nnum = -1;
    double x, y;
    // Material points in the region.
    for (int j = 0; j <= (n_y - 1); ++j)
    {
        for (int i = 0; i <= (n_x - 1); ++i)
        {
            nnum += 1;
            x = i * ds;
            y = j * ds;

            coord[nnum][0] = (R + y) * cos(x / R) + x_c;
            coord[nnum][1] = (R + gamma + y) * sin(x / R) + y_c;
        }
    }

    // Initialize springs
    // Determination of material points inside the horizon of each material point
    std::vector<int> numfam(totnode, 0);
    std::map<Edge, EdgeProp, EdgeComp> mesh;
    for (int i = 0; i < totnode; ++i)
    {
        for (int j = 0; j < totnode; ++j)
        {
            const double idist = sqrt(pow((coord[j][0] - coord[i][0]), 2) + pow((coord[j][1] - coord[i][1]), 2));
            if (i != j)
            {
                if (idist <= delta)
                {
                    numfam[i] += 1;
                    Edge e = std::make_pair(i, j);
                    sort_edge(e);

                    // Determine initial failure of the bond.
                    double fail = 1.0;
                    EdgeProp prop = std::make_pair(idist, fail);
                    mesh[e] = prop;
                }
            }
        }
    }
    const int num_springs = static_cast<int>(mesh.size());

    // Step 1:  Write out the vertex information
    std::fstream vertex_stream;
    vertex_stream.open("shell2d.vertex", std::fstream::out);
    vertex_stream.setf(std::ios_base::scientific);
    vertex_stream.precision(12);

    // first line is the number of vertices in the file
    vertex_stream << totnode << "\n";

    // remaining lines are the initial coordinates of each vertex
    for (int i = 0; i < totnode; ++i)
    {
        vertex_stream << coord[i][0] << "\t" << coord[i][1] << "\n";
    }
    vertex_stream.close();

    // Step 2: Write out the link information (including connectivity and
    // material parameters).
    std::fstream spring_stream;
    spring_stream.open("shell2d.spring", std::fstream::out);
    spring_stream.setf(std::ios_base::scientific);
    spring_stream.precision(12);

    // first line is the number of edges in the file
    spring_stream << num_springs << "\n";

    // remaining lines are the edges in the mesh
    const double K = 0.0;
    const int force_fcn_idx = 0;

    for (std::map<Edge, EdgeProp, EdgeComp>::const_iterator it = mesh.begin(); it != mesh.end(); ++it)
    {
        const Edge& e = it->first;
        const EdgeProp& prop = it->second;

        const int& idx_master = e.first;
        const int& idx_slave = e.second;
        const double& rest_length = prop.first;
        const double& fail = prop.second;
        const double vol_master = vol;
        const double vol_slave = vol;
        spring_stream << idx_master << " " << idx_slave << " " << K << " " << rest_length << " " << force_fcn_idx << " "
                      << vol_master << " " << vol_slave << " " << fail << " " << scr0 << "\n";
    }
    spring_stream.close();
}

File: examples/IB/explicit/ex1/generate_shell2d.m
%% ---------------------------------------------------------------------
%%
%% Copyright (c) 2014 - 2019 by the IBAMR developers
%% All rights reserved.
%%
%% This file is part of IBAMR.
%%
%% IBAMR is free software and is distributed under the 3-clause BSD
%% license. The full text of the license can be found in the file
%% COPYRIGHT at the top level directory of IBAMR.
%%
%% ---------------------------------------------------------------------

function generate_shell2d(N)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Problem parameters
w = 0.0625;
beta  = 0.35;
alpha = 0.25^2/beta;

A = pi*alpha*beta;      % area of ellipse
R = sqrt(A/pi);         % radius of disc with equivalent area as the ellipse
perim = 2*pi*R;         % perimeter of the equivalent disc

dx = 1/N;
dx_64 = 1/64;

num_node_circum = (dx_64/dx)*ceil(perim/(dx_64/3)/4)*4;
ds = perim/num_node_circum;

num_node_radial = (dx_64/dx)*ceil(w/(dx_64/3)/4)*4+1;
dr = w/num_node_radial;

num_node = num_node_radial*num_node_circum;

kappa = 1.0/w;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Step 1: Write out the vertex information
vertex_fid = fopen(['shell2d_' num2str(N) '.vertex'], 'w');

% first line is the number of vertices in the file
fprintf(vertex_fid, '%d\n', num_node);

% remaining lines are the initial coordinates of each vertex
for k = 0:num_node_radial-1
  for l = 0:num_node_circum-1
    theta = 2.0*pi*l/num_node_circum;
    r = k*w/(num_node_radial-1);
    X(1) = 0.5 + (alpha + r)*cos(theta);
    X(2) = 0.5 + (beta  + r)*sin(theta);
    fprintf(vertex_fid, '%1.16e %1.16e\n', X(1), X(2));
  end %for
end %for

fclose(vertex_fid);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Step 2: Write out the link information (including connectivity and
% material parameters).
spring_fid = fopen(['shell2d_' num2str(N) '.spring'], 'w');

% first line is the number of edges in the file
fprintf(spring_fid, '%d\n', num_node_circum*num_node_radial);

% remaining lines are the edges in the mesh
for k = 0:num_node_radial-1
  for l = 0:num_node_circum-1
    current_idx = l;
    next_idx    = current_idx+1;
    if (l == num_node_circum-1)
      next_idx = 0;
    end %if
    current_idx = current_idx + k*num_node_circum;
    next_idx    = next_idx    + k*num_node_circum;
    
    K = kappa*dr/ds;
    rest_length = 0.0;
    
    fprintf(spring_fid, '%6d %6d %1.16e %1.16e\n', current_idx, next_idx, ...
            K, rest_length);
  end %for
end %for 

fclose(spring_fid);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Step 1: Write out the vertex information
vertex_fid = fopen(['shell2d_radial_' num2str(N) '.vertex'], 'w');

% first line is the number of vertices in the file
fprintf(vertex_fid, '%d\n', num_node);

% remaining lines are the initial coordinates of each vertex
for k = 0:num_node_radial-1
  for l = 0:num_node_circum-1
    theta = 2.0*pi*l/num_node_circum;
    r = k*w/(num_node_radial-1);
    X(1) = 0.5 + (alpha + r)*cos(theta);
    X(2) = 0.5 + (beta  + r)*sin(theta);
    fprintf(vertex_fid, '%1.16e %1.16e\n', X(1), X(2));
  end %for
end %for

fclose(vertex_fid);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Step 2: Write out the link information (including connectivity and
% material parameters).
spring_fid = fopen(['shell2d_radial_' num2str(N) '.spring'], 'w');

% first line is the number of edges in the file
fprintf(spring_fid, '%d\n', num_node_circum*(num_node_radial-1));

% remaining lines are the edges in the mesh
for k = 0:num_node_radial-2
  for l = 0:num_node_circum-1
    current_idx = l+ k   *num_node_circum;
    next_idx    = l+(k+1)*num_node_circum;
    
    K = kappa*ds/dr;
    rest_length = 0.0;
    
    fprintf(spring_fid, '%6d %6d %1.16e %1.16e\n', current_idx, next_idx, ...
            K, rest_length);
  end %for
end %for 

fclose(spring_fid);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
